File: LeetCode/299. Bulls and Cows/Solution.py
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        n = len(secret)
        bulls = 0
        cows = 0
        count = [0] * 10
        
        for i in range(n):
            s_digit = int(secret[i])
            g_digit = int(guess[i])
            if s_digit == g_digit:
                bulls += 1
            else:
                # Process secret digit
                if count[s_digit] < 0:
                    cows += 1
                count[s_digit] += 1
                # Process guess digit
                if count[g_digit] > 0:
                    cows += 1
                count[g_digit] -= 1
        
        return f"{bulls}A{cows}B"

File: LeetCode/299. Bulls and Cows/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
    public:
    // Method 1
        // std::string getHint(std::string secret, std::string guess) {
        //     int n = secret.length();
        //     int bulls = 0, cows = 0;
        //     std::vector<int> count(10, 0);

        //     for(int i = 0; i < n; ++i) {
        //         int s = secret[i] - '0';
        //         int g = guess[i] - '0';

        //         if(s == g) {
        //             bulls++;
        //         } else {
        //             if(count[s] < 0) cows++;
        //             ++count[s];
        //             if(count[g] > 0) cows++;
        //             --count[g];
        //         }
        //     }

        //     return std::to_string(bulls) + "A" + std::to_string(cows) + "B";
        // }

        // Method 2 
        string getHint(string secret, string guess) {
            int n = secret.length();
            int bulls = 0, cows = 0;
            int total_matches = 0;

            vector<int> freq_s(10, 0);
            vector<int> freq_g(10, 0);

            for(int i = 0; i < n; ++i) {
                char s = secret[i];
                char g = guess[i];

                if(s == g) {
                    bulls++;
                }

                ++freq_s[s - '0'];
                ++freq_g[g - '0'];
            }

            for (int d = 0; d < 10; ++d) {
                total_matches += min(freq_s[d], freq_g[d]);
            }

            int cows = total_matches - bulls;
            return to_string(bulls) + "A" + to_string(cows) + "B";
        }


}


